docs/dev_guide.md

```diff
@@ -1,31 +1,166 @@
+   added. (
+   Ref: [Sample test file](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/internal/fs/hns_bucket_test.go))
+3. **Structure tests with Arrange-Act-Assert (AAA):** This helps organize your
+   tests and makes them easier to understand.
+    - **Arrange:** Set up the test data and environment.
+    - **Act:** Run the code you're testing.
+    - **Assert:** Check that the results are what you expect.
+
+### How to write end-to-end tests
+
+End-to-end (e2e) tests are crucial for ensuring the correctness and reliability
+of GCSFuse. These tests verify the complete functionality of GCSFuse by
+interacting with a real Google Cloud Storage bucket. This guide explains how to
+write end-to-end tests for GCSFuse.
+
+1. **Adding tests:**
+    - Navigate to
+      the ["integration_tests"](https://github.com/GoogleCloudPlatform/gcsfuse/tree/master/tools/integration_tests)
+      directory which contains all GCSFuse end-to-end tests.
+    - If you are adding a new feature (e.g., a new flag, a new file operation),
+      create a new package directory for your tests. Add corresponding package
+      name entry to
+      our [test script](https://github.com/GoogleCloudPlatform/gcsfuse/blob/a4fb11ad4d424fcc07727dc18e06561ffd8e2467/tools/integration_tests/run_e2e_tests.sh#L66)
+      so that these tests can be run as part of our CI pipeline.
+      Ref PR: https://github.com/GoogleCloudPlatform/gcsfuse/pull/2144
+    - If you are modifying existing behavior, modify the corresponding tests
+      within the relevant package.
+
+2. **Structure tests with Arrange-Act-Assert (AAA):** This helps organize your
+   tests and makes them easier to understand.
+    - **Arrange:** Set up the test data and environment.
+    - **Act:** Run the code you're testing.
+    - **Assert:** Check that the results are what you expect.
+3. **Run Tests:** GCSFuse end-to-end tests are run using the following command.
+    - To run all tests in the package:
+       ```shell
+       export TEST_PACKAGE_NAME=<Enter package name here. For eg: operations>
+       export TEST_BUCKET_NAME=<Enter your bucket name here.>
+       GODEBUG=asyncpreemptoff=1 CGO_ENABLED=0 go test ./tools/integration_tests/$TEST_PACKAGE_NAME/... -p 1 -short --integrationTest -v --testbucket=$TEST_BUCKET_NAME --timeout=60m 
+       ```
+    - To run a particular test add `-run` flag to the command:
+       ```shell
+       export TEST_NAME=<Enter particular test name you want to run>
+       export TEST_PACKAGE_NAME=<Enter package name here. For eg: operations>
+       export TEST_BUCKET_NAME=<Enter your bucket name here.>
+       GODEBUG=asyncpreemptoff=1 CGO_ENABLED=0 go test ./tools/integration_tests/$TEST_PACKAGE_NAME/... -p 1 -short --integrationTest -v --testbucket=$TEST_BUCKET_NAME --timeout=60m -run $TEST_NAME
+       ```
+4. **Run all tests as pre-submit:** You can run existing GCSFuse end-to-end
+   tests as a pre-submit by adding `execute-integration-tests` label to the pull
```

Comment:
External user will not be able to add any label on the PR. We have to provide `go test` command or steps to run our e2e test script locally.

Reply:
Thanks for catching this. I will change to "Ask GCSFuse team to apply `execute-integration-tests` label.

Reply:
Although they'll not be able to see which test fails on their PR. They won't be able to access fusion link. Better to share command to tests on them self

Reply:
Discussed offline, we won't be sharing the script with the customer due to following reasons:
1.  the tests are huge and there is cost associated with running the tests, expecting customers to run the whole test suite is not right.  
2. Some of our tests rely on the secrets stored in the project's secret manager which will fail for the customer.

So we are making it code reviewer responsibility to share the failure logs with the customer.